node/src/components/linear_chain_sync/operations.rs

```diff
@@ -91,34 +96,96 @@ where
     }
 }
 
+/// Fetches and stores a block header from the network
 async fn fetch_and_store_block_header<REv, I>(
     effect_builder: EffectBuilder<REv>,
     block_hash: BlockHash,
-) -> Result<Box<BlockHeader>, LinearChainSyncError<I>>
+) -> Result<Box<BlockHeader>, FetcherError<BlockHeader, I>>
 where
     REv: From<FetcherRequest<I, BlockHeader>> + From<NetworkInfoRequest<I>> + From<StorageRequest>,
     I: Eq + Debug + Clone + Send + 'static,
 {
     let block_header =
         fetch_retry_forever::<BlockHeader, REv, I>(effect_builder, block_hash).await?;
-    if !effect_builder
+    effect_builder
         .put_block_header_to_storage(block_header.clone())
-        .await
+        .await;
+    Ok(block_header)
+}
+
+fn validate_finality_signatures(
+    block_header: &BlockHeader,
+    trusted_validator_weights: &BTreeMap<PublicKey, U512>,
+    finality_threshold_fraction: Ratio<u64>,
+    block_signatures: &BlockSignatures,
+) -> Result<(), FinalitySignatureError> {
+    // Check the signatures' block hash is the header's block hash
+    let block_hash = block_header.hash();
+    if block_signatures.block_hash != block_hash {
+        return Err(
+            FinalitySignatureError::SignaturesDoNotCorrespondToBlockHeader {
+                block_header: Box::new(block_header.clone()),
+                block_hash: Box::new(block_hash),
+                block_signatures: Box::new(block_signatures.clone()),
+            },
+        );
+    }
+
+    // Cryptographically verify block signatures
+    block_signatures.verify()?;
+
+    // Calculate the weight of the signatures
+    let mut signature_weight: U512 = U512::zero();
+    for (public_key, _) in block_signatures.proofs.iter() {
+        match trusted_validator_weights.get(public_key) {
+            None => {
+                return Err(FinalitySignatureError::BogusValidator {
+                    trusted_validator_weights: trusted_validator_weights.clone(),
+                    block_signatures: Box::new(block_signatures.clone()),
+                    bogus_validator_public_key: Box::new(public_key.clone()),
+                })
+            }
+            Some(validator_weight) => {
+                signature_weight += *validator_weight;
+            }
+        }
+    }
+
+    // Check the finality signatures have sufficient weight
+    let total_weight: U512 = trusted_validator_weights
+        .iter()
+        .map(|(_, weight)| *weight)
+        .sum();
+
+    let lower_bound = finality_threshold_fraction.add(1).div(2);
+    // Verify: signature_weight / total_weight >= lower_bound
+    // Equivalent to the following
+    if signature_weight * U512::from(*lower_bound.denom())
+        < total_weight * U512::from(*lower_bound.numer())
```

Comment:
We are basically assuming here that
* numerator and denominator in the FTT have at most 63 bits, and
* the total weight has at most 448 bits,

so this doesn't overflow. All of that is, of course, true in mainnet (and probably always, in practice), but it would be good to make it explicit somewhere.

Reply:
We could also just promote everything to bigints here too... not 100% on how to do that with U512s.

Reply:
I believe we make some assumptions like this elsewhere, too, and in practice, 63 bits will always be enough. We should just enforce it on initialization somehow.